lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -148,6 +154,19 @@ def fake_metadata_controller(tar_path):
         cm = CacheManager(server_config, make_logger)
         # fetching metadata from metadata.log file and key/value not
         # being there should result in a MetadataError
+        tar_name = source_tarball.name.removesuffix(".tar.xz")
+
+        expected_metaerror = f"A problem occurred processing metadata.log from {source_tarball!s}: \"\'filename {tar_name}/metadata.log not found\'\""
+        monkeypatch.setattr(Tarball, "extract", fake_extract)
+        with pytest.raises(Exception) as exc:
+            cm.create(source_tarball)
+        assert str(exc.value) == expected_metaerror
```

Comment:
Should we be either catching something more specific than `Exception` or verifying that `exc` was an instance of some particular exception class?  (Yes, we confirm that it has the right message, which is important for humans, but knowing that it has the right exception instance would be potentially important for calling code.)

Ditto a line 167.

Reply:
Aren't we going to catch that specific exception through `Exception`, So, shouldn't we try to test the real scenario here?

Reply:
The unit test is supposed to ensure that our CUT behaves as expected.  What do we expect `create()` to do when its behavior is driven by `fake_extract()`?  I doubt it's actually going to raise an `Exception` -- rather, it will raise a subclass of that...so, which subclass do we expect it to raise?  (Put another way, what is the "interface contract" of `create()` in this situation?  Surely, it is more specific than just "raises an exception"?...)  That is what we should be checking for at line 161 and/or in addition to 163.

Reply:
According to my `fake_extract` we are raising `KeyError`. After `KeyError` is raised with the message from `fake_extract` it is caught by `create()`. `create()` catces the ``KeyError` in `Exception` which is then again raised through `MetadataError`. Raised `MetadataError` includes the message that `KeyError` has included in it and that is caught in pytest.raises(Exception). Either I can include a specific catch phrase for the errors that can be raised, like KeyError or TarError but unless we don't have those, they will be caught in  `Exception` block of create() which is again raised as `MetadataError`.

Reply:
The unit test code here is testing `create()`, so it should be checking the behavior of that function (under the various conditions that we expect it might face).  We expect `create()` to raise `MetadataError`, so the unit test should ensure that it receives that exception object (specifically).  In addition, we can check the message associated with it, if we think that that is important (which, given that that is what you've coded, I think we do think that 🙂).

That said, looking at the body of `Tarball.extract()`, the only exception that it will raise is `MetadataError`, so `fake_extract()` is not a good representation of that function.

Therefore, I recommend two changes:  `fake_extract()` should be changed to raise `MetadataError` instead of `KeyError`, and the `pytest.raises()` call should specify `MetadataError` instead of `Exception`.

If  you agree with that recommendation, then the `fake_extract_tarfile()` needs some consideration:  we should really make the same change to that, for the same reason, in which case it will become identical to `fake_extract()`, in which case it is redundant, and we should just remove it and the code which uses it.

----

Now, this prompts a look at the other mocks which this test is employing, which all replace `Tarball._get_metadata()`.  Since `CacheManager.create()` calls `Tarball._get_metadata()` which in turn calls `Tarball.extract()`, the fact that you're mocking different levels of this call tree at different times calls into question what your definition of "unit" is, here.  That is, does the unit encompass `CacheManager` and everything it calls (in which case, you should really be mocking `tarfile.open()` instead of `Tarball.extract()`) or is it limited to the `CacheManager.create()` method itself (in which case you should be mocking `Tarball._get_metadata()` in every case and control should never reach `Tarball.extract()`)?

And, as long as I'm asking awkward questions, why is a `CacheManager` method calling an _internal_ method of the `Tarball` class?  (I.e., why does the name of `_get_metadata()` begin with an underscore if it is referenced outside the `Tarball` class?)

----

As long as we're talking about unit testing, I encourage you to look at the Cobertura coverage report for your PR (e.g., for [`cache_manager.py`](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3193/22/cobertura/server/cache_manager_py/)).  You've modified `Tarball.extract()`, but there is no unit test coverage for it.  You've modified `Tarball._get_metadata()`, but it's only half covered.  And, in `CacheManager.create()`, there's coverage for only one of the two `except` clauses that you added.

We can argue that `Tarball.extract()` is too simple to bother with unit testing for it, but the counter-argument is that that makes it very easy to test....  🙂  How we get full coverage of `CacheManager.create()` will depend on what you determine to be the "unit", here.

Reply:
In terms of `Tarball.extract`, I could definitely see a unit test case with a mocked  `tarfile.extractfile` raising `KeyError`, which I believe is what should happen when you try to extract a non-existent file from the tar; the unit test would be proving that it (or actually any exception from `tarfile`) is reported *outside* as a `MetadataError`.

But as Webb says, when mocking `Tarball.extract`... we always expect `MetadataError` from that. So while I like your `KeyError` and `TarError` cases, Webb's correct that these need to be tested at a lower level by mocking `tarfile` directly instead of `Tarball.extract`.

Reply:
> while I like your `KeyError` and `TarError` cases, Webb's correct that these need to be tested at a lower level by mocking `tarfile` directly instead of `Tarball.extract`.

Or...we could define the "units" to be `CacheManager.create()` and `Tarball.extract()`, respectively, and test them separately.  With that approach, we would need (say) two test scenarios for `Tarball.extract()` (proving that it handles each of `KeyError` and `TarError` correctly), and then only a single test scenario for `CacheManager.create()` (proving that it handles `MetadataError` appropriately).

(My instinct is that, because we went to the trouble of defining separate classes for `CacheManager` and `Tarball`, we should be testing them separately:  we should try to keep the unit testing decoupled for all the same reasons that we try to keep the individual classes decoupled.)

Reply:
> Therefore, I recommend two changes: `fake_extract()` should be changed to raise `MetadataError` instead of `KeyError`, and the `pytest.raises()` call should specify `MetadataError` instead of `Exception`.

I have fixed this error and removed the Redundant `fake_extract_tarfile()` Error.